### ISSUE TITLE
why do you iterate height before width for conv_index?

### ISSUE THREAD

**Issue by dhruvbajpai** [allanzelener/YAD2K#19]
on line number 136 in keras_yolo inside the function yolo_head
you say that 
**# Note: YOLO iterates over height index before width index.**

You calculate pred_xy(used for IOU calculations in yolo_loss) by calling yolo_head.
In yolo_head you calculate pred_xy by
**adding to the predicted offsets box_xy  and the index values conv_index.**

which is calculated by : 
conv_index = K.transpose(K.stack([conv_width_index,conv_height_index]))

**shouldn't this be** 
conv_index = K.transpose(K.stack([conv_height_index,conv_width_index]))

like for example 
conv[0][2][3] 
should be    **array([[2,3]])**          **instead of** 
**array([[3,2]])**    <-------**THIS IS IMPLEMENTED CURRENTLY**

**Comment by allanzelener** [comment]
I don't understand, the current code already is as you say it should be. 

The reason for that note is that we want conv_index to be pairs of indices where the first index is iterated before the second index. I.e., we want:
```python
for j in range(15): # Note j here instead of the usual i
  for i in range(15): # i, the height index, is iterated over first
    print(i, j)
```
In the original Darknet this is implemented as a for loop like this, here we have to use vectorized code that is less readable. If I were to reimplement this from scratch and not allow importing Darknet models then I would make this simpler.

**Comment by dhruvbajpai** [comment]
Sorry, my bad.
What I meant is that currently it is: 
conv_index = K.transpose(K.stack([conv_height_index,conv_width_index]))
which gives e.g.
conv_index[0][2][3] =  array([[3,2]])

shouldnt it be
**conv_index = K.transpose(K.stack([conv_width_index,conv_height_index]))**
which would give ----->>  conv_index[0][2][3] =  array([[2,3]])
like in general ------------>> conv_index[k][x][y] = array([[X,Y]])

as the **value of conv_index** added here:
box_xy = (box_xy + conv_index) / conv_dims

is supposed to be the **x,y coordinate value respectively** representing a grid cell added to box_xy in the order x then y. so for conv_index[0][5][4] we should be getting array([[5,4]]) and not array([[4,5]]).

**Comment by allanzelener** [comment]
Yes, box_xy are the predicted (x, y) coordinates for the grid cell at each position (i, j). Note however that (x, y) refers to (width, height) dimensions while (i, j) refers to (height, width) dimensions due to the ordering defined by Keras/Tensorflow.  So at spatial position (i, j) we do in fact want the adjusted position  (x', y') to be (x + j, y + i).

E.g., if x = 0.3, y = 0.7, i = 4, j = 5, and conv_height = conv_width = 13 then the predicted center of the bounding box is at (5.3 / 13, 4.7 / 13) or roughly 41% right and 36% down from the top left corner of the image. 

If I were to design this from scratch I'd make it so that the model predicts (y, x) so you could just add (i, j) but doing this would break compatibility with models trained in Darknet.

I reopened the issue because I think the code isn't very readable here and I found an issue with non-square images, i.e. when  conv_height != conv_width. In that case it's necessary to do something like `conv_index.reshape([width, height, 2]).transpose([1,0,2])` instead of just `conv_index.reshape([height, width, 2])`.
